[jit] move private implementation out of `jit/__init__.py`

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40902 [jit] move script-related implementation out of torch/jit/__init__.py
* **#40807 [jit] move private implementation out of `jit/__init__.py`**

We pack a lot of logic into `jit/__init__.py`, making it unclear to
developers and users which parts of our API are public vs. internal. This
is one in a series of PRs intended to pull implementation out into
separate files, and leave `__init__.py` as a place to register the
public API.

This PR moves all the tracing-related stuff out, and fixes other spots up
as necessary. Followups will move other core APIs out.

The desired end-state is that we conform to the relevant rules in [PEP 8](https://www.python.org/dev/peps/pep-0008/#public-and-internal-interfaces). In particular:
- Internal implementation goes in modules prefixed by `_`. 
- `__init__.py` exposes a public API from these private modules, and nothing more.
- We set `__all__` appropriately to declare our public API.
- All use of JIT-internal functionality outside the JIT are removed (in particular, ONNX is relying on a number internal APIs). Since they will need to be imported explicitly, it will be easier to catch new uses of internal APIs in review.

Differential Revision: [D22320645](https://our.internmc.facebook.com/intern/diff/D22320645)

suo committed:
[jit] move private implementation out of `jit/__init__.py`

We pack a lot of logic into `jit/__init__.py`, making it unclear to
developers and users which parts of our API are public or private. This
is one in a series of PRs intended to pull implementation out into
separate files, and leave `__init__.py` as a place to register the
public API.

I'm going to try to make us conform to the relevant rules in [PEP 8](https://www.python.org/dev/peps/pep-0008/#public-and-internal-interfaces).

[ghstack-poisoned]

suo committed:
Update on "[jit] move private implementation out of `jit/__init__.py`"


We pack a lot of logic into `jit/__init__.py`, making it unclear to
developers and users which parts of our API are public or private. This
is one in a series of PRs intended to pull implementation out into
separate files, and leave `__init__.py` as a place to register the
public API.

I'm going to try to make us conform to the relevant rules in [PEP 8](https://www.python.org/dev/peps/pep-0008/#public-and-internal-interfaces).

[ghstack-poisoned]

suo committed:
Update on "[jit] move private implementation out of `jit/__init__.py`"


We pack a lot of logic into `jit/__init__.py`, making it unclear to
developers and users which parts of our API are public vs. internal. This
is one in a series of PRs intended to pull implementation out into
separate files, and leave `__init__.py` as a place to register the
public API.

This PR moves all the tracing-related stuff out, and fixes other spots up
as necessary. Followups will move other core APIs out.

The desired end-state is that we conform to the relevant rules in [PEP 8](https://www.python.org/dev/peps/pep-0008/#public-and-internal-interfaces). In particular:
- Internal implementation goes in modules prefixed by `_`. 
- `__init__.py` exposes a public API from these private modules, and nothing more.
- We set `__all__` appropriately to declare our public API.
- All use of JIT-internal functionality outside the JIT are removed (in particular, ONNX is relying on a number internal APIs). Since they will need to be imported explicitly, it will be easier to catch new uses of internal APIs in review.

[ghstack-poisoned]

suo committed:
Update on "[jit] move private implementation out of `jit/__init__.py`"


We pack a lot of logic into `jit/__init__.py`, making it unclear to
developers and users which parts of our API are public vs. internal. This
is one in a series of PRs intended to pull implementation out into
separate files, and leave `__init__.py` as a place to register the
public API.

This PR moves all the tracing-related stuff out, and fixes other spots up
as necessary. Followups will move other core APIs out.

The desired end-state is that we conform to the relevant rules in [PEP 8](https://www.python.org/dev/peps/pep-0008/#public-and-internal-interfaces). In particular:
- Internal implementation goes in modules prefixed by `_`. 
- `__init__.py` exposes a public API from these private modules, and nothing more.
- We set `__all__` appropriately to declare our public API.
- All use of JIT-internal functionality outside the JIT are removed (in particular, ONNX is relying on a number internal APIs). Since they will need to be imported explicitly, it will be easier to catch new uses of internal APIs in review.

Differential Revision: [D22320645](https://our.internmc.facebook.com/intern/diff/D22320645)

[ghstack-poisoned]

suo committed:
Update on "[jit] move private implementation out of `jit/__init__.py`"


We pack a lot of logic into `jit/__init__.py`, making it unclear to
developers and users which parts of our API are public vs. internal. This
is one in a series of PRs intended to pull implementation out into
separate files, and leave `__init__.py` as a place to register the
public API.

This PR moves all the tracing-related stuff out, and fixes other spots up
as necessary. Followups will move other core APIs out.

The desired end-state is that we conform to the relevant rules in [PEP 8](https://www.python.org/dev/peps/pep-0008/#public-and-internal-interfaces). In particular:
- Internal implementation goes in modules prefixed by `_`. 
- `__init__.py` exposes a public API from these private modules, and nothing more.
- We set `__all__` appropriately to declare our public API.
- All use of JIT-internal functionality outside the JIT are removed (in particular, ONNX is relying on a number internal APIs). Since they will need to be imported explicitly, it will be easier to catch new uses of internal APIs in review.

Differential Revision: [D22320645](https://our.internmc.facebook.com/intern/diff/D22320645)

[ghstack-poisoned]

suo committed:
Update on "[jit] move private implementation out of `jit/__init__.py`"


We pack a lot of logic into `jit/__init__.py`, making it unclear to
developers and users which parts of our API are public vs. internal. This
is one in a series of PRs intended to pull implementation out into
separate files, and leave `__init__.py` as a place to register the
public API.

This PR moves all the tracing-related stuff out, and fixes other spots up
as necessary. Followups will move other core APIs out.

The desired end-state is that we conform to the relevant rules in [PEP 8](https://www.python.org/dev/peps/pep-0008/#public-and-internal-interfaces). In particular:
- Internal implementation goes in modules prefixed by `_`. 
- `__init__.py` exposes a public API from these private modules, and nothing more.
- We set `__all__` appropriately to declare our public API.
- All use of JIT-internal functionality outside the JIT are removed (in particular, ONNX is relying on a number internal APIs). Since they will need to be imported explicitly, it will be easier to catch new uses of internal APIs in review.

Differential Revision: [D22320645](https://our.internmc.facebook.com/intern/diff/D22320645)

[ghstack-poisoned]

eellison commented on torch/jit/_script.py:
```diff
@@ -0,0 +1,169 @@
+    def Attribute(value, type):
+        return value
+
+
```
reason why `script` is not defined in this file ?

suo replied:
I was limiting this first PR to the things necessary to move tracing out, just to keep things manageable. I can either roll the scripting-related changes into this one or do it in a followup.

eellison commented on torch/jit/_script.py:
```diff
@@ -0,0 +1,169 @@
+    # did nothing __getattr__ would not be called. Instead we'd get nn.Module.forward
+    # which always throws an exception.
+
+    class ScriptModule(with_metaclass(ScriptMeta, Module)):
```
should `RecursiveScriptModule` / `ScriptMeta` / `script` `script_method` be in this file ?

suo replied:
see above—I can stack some more stuff on top so that we can get a better sense of what the end result looks like

eellison commented on torch/jit/_state.py:
```diff
@@ -0,0 +1,57 @@
+
+def enable():
+    _enabled.enabled = True
+
```
Maybe add `_jit_caching_layer` and `_jit_function_overload_caching` here ?

suo replied:
same as above

eellison reviewed (commented):
LGTM! That file was huge so this will be a nice refactor. A few changes i think we should also do. 

Happy to accept this but i think it might be good to get one more eye on this.

suo committed:
Update on "[jit] move private implementation out of `jit/__init__.py`"


We pack a lot of logic into `jit/__init__.py`, making it unclear to
developers and users which parts of our API are public vs. internal. This
is one in a series of PRs intended to pull implementation out into
separate files, and leave `__init__.py` as a place to register the
public API.

This PR moves all the tracing-related stuff out, and fixes other spots up
as necessary. Followups will move other core APIs out.

The desired end-state is that we conform to the relevant rules in [PEP 8](https://www.python.org/dev/peps/pep-0008/#public-and-internal-interfaces). In particular:
- Internal implementation goes in modules prefixed by `_`. 
- `__init__.py` exposes a public API from these private modules, and nothing more.
- We set `__all__` appropriately to declare our public API.
- All use of JIT-internal functionality outside the JIT are removed (in particular, ONNX is relying on a number internal APIs). Since they will need to be imported explicitly, it will be easier to catch new uses of internal APIs in review.

Differential Revision: [D22320645](https://our.internmc.facebook.com/intern/diff/D22320645)

[ghstack-poisoned]

suo committed:
Update on "[jit] move private implementation out of `jit/__init__.py`"


We pack a lot of logic into `jit/__init__.py`, making it unclear to
developers and users which parts of our API are public vs. internal. This
is one in a series of PRs intended to pull implementation out into
separate files, and leave `__init__.py` as a place to register the
public API.

This PR moves all the tracing-related stuff out, and fixes other spots up
as necessary. Followups will move other core APIs out.

The desired end-state is that we conform to the relevant rules in [PEP 8](https://www.python.org/dev/peps/pep-0008/#public-and-internal-interfaces). In particular:
- Internal implementation goes in modules prefixed by `_`. 
- `__init__.py` exposes a public API from these private modules, and nothing more.
- We set `__all__` appropriately to declare our public API.
- All use of JIT-internal functionality outside the JIT are removed (in particular, ONNX is relying on a number internal APIs). Since they will need to be imported explicitly, it will be easier to catch new uses of internal APIs in review.

Differential Revision: [D22320645](https://our.internmc.facebook.com/intern/diff/D22320645)

[ghstack-poisoned]

suo committed:
Update on "[jit] move private implementation out of `jit/__init__.py`"


We pack a lot of logic into `jit/__init__.py`, making it unclear to
developers and users which parts of our API are public vs. internal. This
is one in a series of PRs intended to pull implementation out into
separate files, and leave `__init__.py` as a place to register the
public API.

This PR moves all the tracing-related stuff out, and fixes other spots up
as necessary. Followups will move other core APIs out.

The desired end-state is that we conform to the relevant rules in [PEP 8](https://www.python.org/dev/peps/pep-0008/#public-and-internal-interfaces). In particular:
- Internal implementation goes in modules prefixed by `_`. 
- `__init__.py` exposes a public API from these private modules, and nothing more.
- We set `__all__` appropriately to declare our public API.
- All use of JIT-internal functionality outside the JIT are removed (in particular, ONNX is relying on a number internal APIs). Since they will need to be imported explicitly, it will be easier to catch new uses of internal APIs in review.

Differential Revision: [D22320645](https://our.internmc.facebook.com/intern/diff/D22320645)

[ghstack-poisoned]
